src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,219 +287,397 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
 
-    protected virtual Task<ExecutionResult> ExecuteRequestAsync(
-        GraphQLRequest gqlRequest,
-        IDictionary<string, object> userContext,
-        IDocumentExecuter<TSchema> executer,
-        IServiceProvider requestServices,
-        CancellationToken token)
-        => executer.ExecuteAsync(new ExecutionOptions
+        if (socket.SubProtocol != selectedProtocol)
         {
-            Query = gqlRequest.Query,
-            OperationName = gqlRequest.OperationName,
-            Variables = gqlRequest.Variables,
-            Extensions = gqlRequest.Extensions,
-            UserContext = userContext,
-            RequestServices = requestServices,
-            CancellationToken = token
-        });
+            await socket.CloseAsync(
+                WebSocketCloseStatus.ProtocolError,
+                $"Invalid sub-protocol; expected '{selectedProtocol}'",
+                context.RequestAborted);
+            return;
+        }
 
-    protected virtual CancellationToken GetCancellationToken(HttpContext context) => context.RequestAborted;
+        // Prepare user context
+        var userContext = await BuildUserContextAsync(context);
+        // Connect, then wait until the websocket has disconnected (and all subscriptions ended)
+        await selectedHandler.ExecuteAsync(context, socket, selectedProtocol, userContext);
+    }
 
-    protected virtual Task RequestExecutingAsync(GraphQLRequest request, int? indexInBatch = null)
+    *****************************/
+
+    /// <summary>
+    /// Writes a '401 Access denied.' message to the output when the user is not authenticated.
+    /// </summary>
+    protected virtual Task HandleNotAuthenticatedAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.Unauthorized, new AccessDeniedError("schema"));
+
+    /// <summary>
+    /// Writes a '401 Access denied.' message to the output when the user fails the role checks.
+    /// </summary>
+    protected virtual Task HandleNotAuthorizedRoleAsync(HttpContext context, RequestDelegate next)
+        => WriteErrorResponseAsync(context, HttpStatusCode.Unauthorized, new AccessDeniedError("schema") { RolesRequired = Options.AuthorizedRoles });
+
+    /// <summary>
+    /// Writes a '401 Access denied.' message to the output when the user fails the policy check.
+    /// </summary>
+    protected virtual Task HandleNotAuthorizedPolicyAsync(HttpContext context, RequestDelegate next, AuthorizationResult authorizationResult)
+        => WriteErrorResponseAsync(context, HttpStatusCode.Unauthorized, new AccessDeniedError("schema") { PolicyRequired = Options.AuthorizedPolicy, PolicyAuthorizationResult = authorizationResult });
+
+    /// <summary>
+    /// Writes a '400 JSON body text could not be parsed.' message to the output.
+    /// Return <see langword="false"/> to rethrow the exception or <see langword="true"/>
+    /// if it has been handled.
```

Comment:
```suggestion
    /// if it has been handled. By default returns <see langword="true"/>.
```

Reply:
```suggestion
    /// if it has been handled.  By default returns <see langword="true"/>.
```

Reply:
Note that in my future PRs I will follow default style conventions regarding whitespaces which are used in most OSS projects that I have seen.